### PR TITLE
Dl text frame iterator prototypes

### PR DESCRIPTION
**Pull request by flar**
This is a draft pull request to demonstrate 3 different ways of iterating over all TextFrame objects in a DisplayList (with recursion).

The Iterate method tracks transform operations as it goes so it can report them to the Iterator.

The Iterate2 method relies on the Builder to have stashed the matrices in the DL op records so it can just focus on finding the required records and calling the Iterator on them.

The benchmark also has an example of a custom Dispatcher that tracks all of the matrix operations and calls an iterator whenever a DrawTextFrame operation is dispatched. This method requires no changes to the DL mechanism, but has additional overhead.

### CODE REVIEW

**Comment by flar** [comment]
The benchmark results for the 3 different operations:

| Benchmark                                                                |  Time             | CPU   | Iterations |
| --- | --- | --- | --- |
BM_DisplayListIterateTextFrames/kDefault                            |  0.395 us     |   0.394 us    |  1808963 |
BM_DisplayListIterateTextFrames2/kDefault                          |   0.168 us     |   0.167 us     | 4136725 |
BM_DisplayListDispatchTextFrames/kDefault                        |     0.776 us    |    0.774 us   |    900426 |

**Review by jonahwilliams — COMMENTED** [review]
How does tracking the matrix per op work with nested display lists?

File: display_list/display_list.cc [hunk 1]
```diff
@@ -148,6 +149,121 @@ void DisplayList::Dispatch(DlOpReceiver& receiver,
...
+    switch (op->type) {
+      case DisplayListOpType::kDrawTextFrame: {
+        auto text_frame_op = static_cast<const DrawTextFrameOp*>(op);
+        iterator(*text_frame_op->text_frame,              //
+                 initial_matrix * text_frame_op->matrix,  //
+                 text_frame_op->x, text_frame_op->y);
+        break;
+      }
+
+      case DisplayListOpType::kDrawDisplayList: {
```
Comment (jonahwilliams) [inline display_list/display_list.cc:254]: Ahh, so we track the matrix per op and then multiple while dispatching. Seems solid! 👍

**Comment by flar** [comment]
If I implement the Dispatcher to just hold a Matrix directly (instead of indirectly via DlMatrixClipTracker), it gets a little better in performance:

| Benchmark                                                            |     Time     |        CPU  | Iterations |
| --- | --- | --- | --- |
| BM_DisplayListIterateTextFrames/kDefault        |                      0.388 us     |   0.387 us   |   1806568 |
| BM_DisplayListIterateTextFrames2/kDefault       |                      0.167 us    |    0.167 us   |   4191140 |
| BM_DisplayListDispatchTextFrames/kDefault      |                       0.621 us  |      0.619 us  |    1131953 |

**Comment by flar** [comment]
Given how quick the entire process is, I think the non-invasive Dispatcher object I created in the benchmark file would probably disappear into the noise of managing or filling the atlas and would also provide a basis for any other future preview work that Impeller needs to do on a scene.

**Comment by jonahwilliams** [comment]
That SGTM.

**Comment by jonahwilliams** [comment]
I pulled the dispatcher approach into https://github.com/flutter/engine/pull/52035 and it works nicely!

**Comment by flar** [comment]
Archiving PR now that the solution was decided on.

**Pull request closed without merging**